plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,567 @@
+algoltek_device_bot(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 5;
+	buf->data[1] = ALGOLTEK_BOT;
+	fu_memwrite_uint16(buf->data + 2, address, G_BIG_ENDIAN);
+	buf->data[10] = ~fu_sum8(buf->data, buf->len) + 1;
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   ALGOLTEK_BOT,
+					   0, /* value */
+					   0, /* index */
+					   buf->data,
+					   buf->data[0],
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to contact device: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_ers(FuAlgoltekUsbDevice *self, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 2, 0x0);
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   ALGOLTEK_ERS,
+					   0, /* value */
+					   0, /* index */
+					   buf->data,
+					   buf->len,
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to contact device: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_status_check(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) update_status_array = g_byte_array_new();
+	guint retryTimes = 0;
+	guint16 update_status;
+Retry:
+	update_status_array = algoltek_device_rdr(self, AG_UPDATE_STATUS, error);
+	if (update_status_array == NULL)
+		return FALSE;
+	update_status = fu_algoltek_readout_value(update_status_array);
+	switch (update_status) {
+	case AG_UPDATE_PASS:
+		break;
+	case AG_UPDATE_FAIL:
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Update failure.");
+		return FALSE;
+	default:
+		retryTimes++;
+		if (retryTimes < 10)
+			goto Retry;
+		else
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_wrf(FuAlgoltekUsbDevice *self,
+		    GByteArray *firmware_data,
+		    FuProgress *progress,
+		    GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) transfer_parameter = g_byte_array_new();
+	g_autoptr(GByteArray) part_firmware = g_byte_array_new();
+	guint rest_transfer_data_size = firmware_data->len;
+	guint data_addr = 0;
+	guint start_index = 0;
+	guint max_packet_size = 64;
+	guint transfer_data_size = 0;
+	guint count_check = 0;
+	guint16 value;
+	guint16 index;
+
+	while (rest_transfer_data_size > 0) {
+		count_check++;
+		data_addr = start_index;
```

Comment:
fu_chunk_array_new will help a lot here

Reply:
Codes will be revised according to the suggestion.

Reply:
Updated! Use `fu_chunk_array_new( )` instead of while loop.